[IMP] scss: add section, improve preexisting content (#3093)
odoo/documentation: content/contributing/development/ui/scss_optimization.rst

.. warning::
   The fact that a class achieve the desired visual result doesn't necessarily mean that's the right one for the job. Be aware of classes triggering JS behaviors, for example.

.. warning::
   Take care about classes semantic. To apply a **button class** to a **title** is not just semantically wrong, it may also lead to migration issues and visual inconsistencies.

The following sections describe tips to strip-down SCSS lines **when custom-code is the only way to go**.

Browsers defaults
=================

By default, each browser renders content using an *user agent stylesheet*.
To overcome inconsistencies, some of these rules are overridden by `Bootstrap Reboot <https://getbootstrap.com/docs/5.1/content/reboot/>`_.

At this stage all "browser-specific-decoration" rules have been stripped away, but a big chunk of rules defining basic layout information are maintained (or reinforced by *Reboot* for consistency reason).

You can rely on these rules.

> For example, apply `display: block;` to a `<div/>` is normally not necessary.
> 
> .. code-block:: css
> 
>    div.myElement {
>       display: block;
>       /* not needed 99% of the time */
>    }

Reviewer: ```suggestion
.. example::

   Apply `display: block;` to a `<div/>` is normally not necessary.

   .. code-block:: CSS

      div.myElement {
         display: block;
         /* not needed 99% of the time */
      }
```

For the same reason, sometimes you may opt to switch html tag rather than adding a new CSS rule.

.. code-block:: css

   span.myElement {
      display: block;
      /* replace <span> with <div> instead
         to get 'display: block' by default */
   }

Follows a non-comprehensive list of default rules:

.. list-table::
   :header-rows: 1

   * - Tag / Attribute
     - Defaults
   * - | `<div/>`, `<section/>`,

>        | `<header/>`, `<footer/>` ...

Reviewer: Why do you force a line break here?

Author: To visually align elements and avoid the table to break the design

     - `display: block`
   * - | `<span/>`, `<a/>`
       | `<em/>`, `<b/>` ...
     - `display: inline`
   * - | `<button/>`, `<label/>`
       | `<output/>`...
     - `display: inline-block`
   * - `<img/>`, `<svg/>`
     - `vertical-align: middle`
   * - `<summary/>`, `[role="button"]`
     - `cursor: pointer;`
   * - `<q/>`
     - | `:before {content: open-quote}`
       | `:after  {content: close-quote}`
   * - ...
     - ...

.. seealso::
   Bootstrap Reboot `on github <https://github.com/twbs/bootstrap/blob/1a6fdfae6b/scss/_reboot.scss>`_

Use tags
========

| It may seems obvious, but the most simple and **consistent** way to "convert" a text to a title is to use a title html tag (`<h1>`, `<h2>`, ...).
| Beside reboot rules, mostly all tags carry decorative style defined by Odoo.

> Dont'

Reviewer: Same suggestion as above.

Author: Can't use `code-tabs` inside `.. example::` (can I ?)

=========
SCSS Tips
=========

There are many ways to lean and simplify SCSS.
The first step is to ensablish if custom code is needed at all.

> Odoo's webclient has been designed to be modular, meaning that (potentially all) classes can be shared across views.
> Check the code before creating a new class. Chances are that there is already a class or a tag doing exaclty what you're looking for.

Reviewer 2: ```suggestion
Odoo's webclient has been designed to be modular, meaning that (potentially all) classes can be shared across views.
Check the code before creating a new class. Chances are that there is already a class or a tag doing exactly what you're looking for.
```

> On top of that, odoo relies on `Bootstrap <https://getbootstrap.com/docs/5.1/getting-started/introduction/>`_ (BS), one of the most complete CSS framework available. The framework has been customised in order to match odoo design (both community and enterprise versions), meaning that you can use any BS class, direclty in odoo and achieve a visual result consistent with our UI.

Reviewer 2: ```suggestion
On top of that, odoo relies on `Bootstrap <https://getbootstrap.com/docs/5.1/getting-started/introduction/>`_ (BS), one of the most complete CSS framework available. The framework has been customized in order to match odoo design (both community and enterprise versions), meaning that you can use any BS class, directly in odoo and achieve a visual result consistent with our UI.
```
And maybe the 3 `odoo` -> `Odoo`

.. warning::

>    The fact that a class achieve the desired visual result doesn't necessarily mean that's the right one for the job. Be aware of classes triggering JS behaviours, for example.

Reviewer 2: ```suggestion
   The fact that a class achieve the desired visual result doesn't necessarily mean that's the right one for the job. Be aware of classes triggering JS behaviors, for example.
```

.. warning::

>    Take care about classes semantic. To apply a **button class** to a **title** is not just sematically wrong, it may also lead to migration issues and visual incostistencies.

Reviewer 2: ```suggestion
   Take care about classes semantic. To apply a **button class** to a **title** is not just semantically wrong, it may also lead to migration issues and visual inconsistencies.
```

The following sections describe tips to strip-down SCSS lines **when custom-code is the only way to go**.

Browsers defaults
=================

By default, each browser renders content using an *user agent stylesheet*.
To overcome inconsistencies, some of these rules are overridden by `Bootstrap Reboot <https://getbootstrap.com/docs/5.1/content/reboot/>`_.

At this stage all "browser-specific-decoration" rules have been stripped away, but a big chunk of rules defining basic layout information are maintained (or reinforced by *Reboot* for consistency reason).

You can rely on these rules.

For example, apply `display: block;` to a `<div/>` is normally not necessary.

.. code-block:: css

   div.myElement {
      display: block;

>       /* not needed 99% of the times */

Reviewer 2: ```suggestion
      /* not needed 99% of the time */
```

         font-weight: bold;
         animation: 1s linear 1s mycustomAnimation;
      }

      .o_module_custom_subtitle {
         display: block;
         font-size: 12px;
         font-weight: bold;
         animation: 2s linear 1s mycustomAnimation;
      }

Do

.. tabs::

   .. code-tab:: html XML

      <h5 class="o_module_custom_title">
         Hello There!
      </h5>

      <div class="o_module_custom_subtitle">
         <b><small>I'm the subtitle.</small></b>
      </div>

   .. code-tab:: css SCSS

      .o_module_custom_title {
         animation: 1s linear 1s mycustomAnimation;
      }

      .o_module_custom_subtitle {
         animation: 2s linear 1s mycustomAnimation;
      }

.. note::
   Beside reducing the amount of code, a modular-design approach (use classes, tags, mixins...) keeps the visual result consistent and easily **maintainable**.

   Following the last example, if odoo titles' design changes, these changes will be applied in the `o_module_custom_title` element too since it's using a the `<h5>` tag.

Utility classes
===============

Our framework defines a multitude of utility classes designed to cover almost all layout/design/interaction needs.

> The simple fact that a class exist justify its use over custom CSS whenever is possible.

Reviewer 2: ```suggestion
The simple fact that a class exists justify its use over custom CSS whenever is possible.
```
not sure

Take the example of `position-relative`. What does is quite self-explicative, but *why* should be used maybe a little less.

.. code-block:: css

   position-relative {
      position: relative !important;
   }

> Since an utility-class is defined, any CSS line with declaration `position: relative` is **potentially** reduntant.

Reviewer 2: ```suggestion
Since an utility-class is defined, any CSS line with declaration `position: relative` is **potentially** redundant.
```
